frame/election-provider-multi-phase/src/unsigned.rs

```diff
@@ -144,49 +161,67 @@ impl<T: Config> Pallet<T> {
 			Self::snapshot().ok_or(MinerError::SnapshotUnAvailable)?;
 		let desired_targets = Self::desired_targets().ok_or(MinerError::SnapshotUnAvailable)?;
 
-		// closures.
+		// now make some helper closures.
 		let cache = helpers::generate_voter_cache::<T>(&voters);
 		let voter_index = helpers::voter_index_fn::<T>(&cache);
 		let target_index = helpers::target_index_fn::<T>(&targets);
 		let voter_at = helpers::voter_at_fn::<T>(&voters);
 		let target_at = helpers::target_at_fn::<T>(&targets);
 		let stake_of = helpers::stake_of_fn::<T>(&voters, &cache);
 
-		let ElectionResult { assignments, winners } = election_result;
+		// Compute the size of a compact solution comprised of the selected arguments.
+		//
+		// This function completes in `O(edges)`; it's expensive, but linear.
+		let encoded_size_of = |assignments: &[IndexAssignmentOf<T>]| {
+			CompactOf::<T>::try_from(assignments).map(|compact| compact.encoded_size())
+		};
+
+		let ElectionResult { mut assignments, winners } = election_result;
+		// Sort the assignments by reversed voter stake. This ensures that we can efficiently
+		// truncate the list.
+		assignments.sort_by_key(|Assignment::<T> { who, .. }| {
+			let stake = cache.get(who).map(|idx| {
+				let (_, stake, _) = voters[*idx];
+				stake
+			}).unwrap_or_default();
+			sp_std::cmp::Reverse(stake)
+		});
+
+		// Reduce (requires round-trip to staked form)
+		assignments = {
+			// convert to staked and reduce.
+			let mut staked = assignment_ratio_to_staked_normalized(assignments, &stake_of)?;
+			sp_npos_elections::reduce(&mut staked);
```

Comment:
did we implicitely assume that `reudce` does not alter the order of assignments? 

Would be good to call `assignments` `sorted_assignments` to clarify this further.

Reply:
well.. you know.. we could also do this: 

1. have assignments as-is
2. make staked assignment 
3. reduce
4. sort (because staked-assignment has stake and you _can_ sort)
5. convert back to to assignments.

Reply:
either way, lets make sure anywhere that assignments need to be sorted the name reflects that and then we can ensure that we are not making a mistake.

Reply:
> did we implicitely assume that reudce does not alter the order of assignments?

Yes, we did. I'll admit that I haven't looked very closely at how `reduce` works, but it seems at a glance to edit the edges of an assignment, as opposed to the sequence of assignments. 

> Would be good to call assignments sorted_assignments

Ok. 8f0e9c560

Reply:
> Yes, we did. I'll admit that I haven't looked very closely at how reduce works, but it seems at a glance to edit the edges of an assignment, as opposed to the sequence of assignments.


I don't recall exactly either, and it seems like the new flow that I recommended is just as good and does not make these footgun assumptions. (`reduce` is a very badly written code by me in a hurry btw and can be refactored at some point)

Reply:
> the new flow that I recommended is just as good and does not make these footgun assumptions.

Yeah, it is. 26726f8 took a while mainly because I wanted to make sure that it wasn't related to the errors which popped up when I first wrote the benchmark in c9f2517, which was under development simultaneously.

Reply:
yeah, looks very good now.